File: src/nap_msg/asr.py
import base64
import hashlib
import hmac
import json
import os
import time
from datetime import datetime
from typing import Optional

import httpx



def _hmac_sha256(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()


def _build_tc3_headers(body: bytes, timestamp: int, secret_id: str, secret_key: str, region: Optional[str]) -> dict:
    date = datetime.utcfromtimestamp(timestamp).strftime("%Y-%m-%d")
    payload_hash = hashlib.sha256(body).hexdigest()

    canonical_request = "\n".join(
        [
            "POST",
            "/",
            "",
            "content-type:application/json; charset=utf-8",
            "host:asr.tencentcloudapi.com",
            "",
            "content-type;host",
            payload_hash,
        ]
    )

    credential_scope = f"{date}/asr/tc3_request"
    string_to_sign = "\n".join(
        [
            "TC3-HMAC-SHA256",
            str(timestamp),
            credential_scope,
            hashlib.sha256(canonical_request.encode("utf-8")).hexdigest(),
        ]
    )

    secret_date = _hmac_sha256(("TC3" + secret_key).encode("utf-8"), date)
    secret_service = _hmac_sha256(secret_date, "asr")
    secret_signing = _hmac_sha256(secret_service, "tc3_request")
    signature = hmac.new(secret_signing, string_to_sign.encode("utf-8"), hashlib.sha256).hexdigest()

    authorization = (
        f"TC3-HMAC-SHA256 Credential={secret_id}/{credential_scope}, "
        f"SignedHeaders=content-type;host, Signature={signature}"
    )

    headers = {
        "Authorization": authorization,
        "Content-Type": "application/json; charset=utf-8",
        "Host": "asr.tencentcloudapi.com",
        "X-TC-Action": "SentenceRecognition",
        "X-TC-Version": "2019-06-14",
        "X-TC-Timestamp": str(timestamp),
        "X-TC-RequestClient": "maid",
    }
    if region:
        headers["X-TC-Region"] = region

    return headers


async def sentence_recognize(
    data: bytes,
    voice_format: str = "mp3",
    eng_service_type: Optional[str] = None,
    project_id: Optional[int] = None,
) -> str:
    if not data:
        raise ValueError("Audio data is empty")

    secret_id = os.getenv("TENCENT_SECRET_ID", "").strip()
    secret_key = os.getenv("TENCENT_SECRET_KEY", "").strip()
    if not secret_id or not secret_key:
        raise RuntimeError("Missing Tencent Cloud credentials: set TENCENT_SECRET_ID and TENCENT_SECRET_KEY")

    region = os.getenv("TENCENT_ASR_REGION", "").strip() or None
    engine = eng_service_type or os.getenv("TENCENT_ASR_ENGINE", "16k_zh").strip()

    payload = {
        "SubServiceType": 2,
        "EngSerViceType": engine,
        "SourceType": 1,
        "VoiceFormat": voice_format,
        "Data": base64.b64encode(data).decode("utf-8"),
    }
    if project_id is not None:
        payload["ProjectId"] = project_id

    body = json.dumps(payload).encode("utf-8")
    ts = int(time.time())
    headers = _build_tc3_headers(body, ts, secret_id, secret_key, region)

    url = "https://asr.tencentcloudapi.com"
    logger.info("Calling Tencent Cloud sentence recognition")

    async with httpx.AsyncClient(timeout=30.0) as client:
        resp = await client.post(url, content=body, headers=headers)
        resp.raise_for_status()
        result = resp.json()

    response = result.get("Response") if isinstance(result, dict) else None
    if not response:
        raise RuntimeError("Tencent Cloud returned invalid structure")

    if "Error" in response:
        err = response["Error"]
        code = err.get("Code")
        msg = err.get("Message")
        raise RuntimeError(f"Tencent Cloud ASR error: {code} - {msg}")

    text = response.get("Result")
    if not text:
        raise RuntimeError("Tencent Cloud did not return recognition result")

    return text

File: src/asr.ts
import { createHash, createHmac } from "node:crypto";
import { type NapcatAsrConfig } from "./types.js";
import { type NapcatLogger } from "./logger.js";

/**
 * Returns true when the ASR config has both required credentials.
 */
export function isAsrEnabled(asr: NapcatAsrConfig | undefined): boolean {
  return Boolean(asr?.secretId?.trim() && asr?.secretKey?.trim());
}

function hmacSha256(key: Buffer, msg: string): Buffer {
  return createHmac("sha256", key).update(msg, "utf8").digest();
}

function buildTc3Headers(
  body: Buffer,
  timestamp: number,
  secretId: string,
  secretKey: string,
  region: string | undefined,
): Record<string, string> {
  const date = new Date(timestamp * 1000).toISOString().slice(0, 10); // YYYY-MM-DD
  const payloadHash = createHash("sha256").update(body).digest("hex");

  const canonicalRequest = [
    "POST",
    "/",
    "",
    "content-type:application/json; charset=utf-8",
    "host:asr.tencentcloudapi.com",
    "",
    "content-type;host",
    payloadHash,
  ].join("\n");

  const credentialScope = `${date}/asr/tc3_request`;
  const stringToSign = [
    "TC3-HMAC-SHA256",
    String(timestamp),
    credentialScope,
    createHash("sha256").update(canonicalRequest, "utf8").digest("hex"),
  ].join("\n");

  const secretDate = hmacSha256(Buffer.from("TC3" + secretKey, "utf8"), date);
  const secretService = hmacSha256(secretDate, "asr");
  const secretSigning = hmacSha256(secretService, "tc3_request");
  const signature = createHmac("sha256", secretSigning)
    .update(stringToSign, "utf8")
    .digest("hex");

  const authorization =
    `TC3-HMAC-SHA256 Credential=${secretId}/${credentialScope}, ` +
    `SignedHeaders=content-type;host, Signature=${signature}`;

  const headers: Record<string, string> = {
    Authorization: authorization,
    "Content-Type": "application/json; charset=utf-8",
    Host: "asr.tencentcloudapi.com",
    "X-TC-Action": "SentenceRecognition",
    "X-TC-Version": "2019-06-14",
    "X-TC-Timestamp": String(timestamp),
    "X-TC-RequestClient": "maid",
  };

  if (region) {
    headers["X-TC-Region"] = region;
  }

  return headers;
}

/**
 * Call Tencent Cloud SentenceRecognition API and return the recognized text.
 *
 * @param audioBytes  Raw audio data (MP3 or other format supported by the engine)
 * @param voiceFormat Audio format string, e.g. "mp3" (default)
 * @param asr         ASR config from channels.napcat.asr in config.json
 * @param log         Optional logger
 */
export async function sentenceRecognize(
  audioBytes: Buffer,
  voiceFormat: string = "mp3",
  asr: NapcatAsrConfig,
  log?: NapcatLogger,
): Promise<string> {
  if (!audioBytes || audioBytes.length === 0) {
    throw new Error("Audio data is empty");
  }

  const secretId = asr.secretId.trim();
  const secretKey = asr.secretKey.trim();
  if (!secretId || !secretKey) {
    throw new Error("ASR credentials not configured (channels.napcat.asr.secretId / secretKey)");
  }

  const region = asr.region?.trim() || undefined;
  const engine = asr.engine?.trim() || "16k_zh";

  const payload = {
    SubServiceType: 2,
    EngSerViceType: engine,
    SourceType: 1,
    VoiceFormat: voiceFormat,
    Data: audioBytes.toString("base64"),
  };

  const body = Buffer.from(JSON.stringify(payload), "utf8");
  const timestamp = Math.floor(Date.now() / 1000);
  const headers = buildTc3Headers(body, timestamp, secretId, secretKey, region);

  log?.info(`Calling Tencent ASR (engine=${engine}, format=${voiceFormat}, bytes=${audioBytes.length})`);

  const url = "https://asr.tencentcloudapi.com";
  const resp = await fetch(url, {
    method: "POST",
    headers,
    body,
    signal: AbortSignal.timeout(30_000),
  });

  if (!resp.ok) {
    throw new Error(`Tencent Cloud ASR HTTP ${resp.status} ${resp.statusText}`);
  }

  const result = (await resp.json()) as Record<string, unknown>;
  const response = result["Response"] as Record<string, unknown> | undefined;
  if (!response) {
    throw new Error("Tencent Cloud ASR returned invalid structure");
  }

  const error = response["Error"] as Record<string, unknown> | undefined;
  if (error) {
    const code = error["Code"];
    const msg = error["Message"];
    throw new Error(`Tencent Cloud ASR error: ${String(code)} - ${String(msg)}`);
  }

  const text = response["Result"];
  if (!text) {
    throw new Error("Tencent Cloud ASR did not return a recognition result");
  }

  const recognized = String(text);
  log?.info(`ASR result: "${recognized}"`);
  return recognized;
}
